src/Workspaces/Core/Portable/Shared/Utilities/AsyncBatchingWorkQueue.cs

```diff
@@ -89,21 +115,7 @@ public void AddWork(IEnumerable<TItem> items)
             lock (_gate)
             {
                 // add our work to the set we'll process in the next batch.
-                _nextBatch.AddRange(items);
-                TryKickOffNextBatchTask();
-            }
-        }
-
-        public void AddWork(ImmutableArray<TItem> items)
-        {
-            // Don't do any more work if we've been asked to shutdown.
-            if (_cancellationToken.IsCancellationRequested)
-                return;
-
-            lock (_gate)
-            {
-                // add our work to the set we'll process in the next batch.
-                _nextBatch.AddRange(items);
+                _addItemsToBatch(batch: _nextBatch, newItems: items);
```

Comment:
>_addItemsToBatch [](start = 16, length = 16)

Calling arbitrary code under a lock is dangerous. I'd prefer a different design.

Reply:
Why do we need an abstraction here anyways?

---
In reply to: [398956530](https://github.com/dotnet/roslyn/pull/42740#discussion_r398956530) [](ancestors = 398956530)

Reply:
yes.  i can do better here.

the idea is to avoid a flurry of work (i.e. 10k workspace-changed notifications) from needing us to buffer and report all 10k of those changes.  but this is much more easily supported by clients just providing something like an IEqualityComparer they can say can be used to dedupe entries if there are dupes.